lib/optimizer/src/Transformer/ServerSideRendering.php

```diff
@@ -620,4 +699,260 @@ private function isAmpExperimentUsed(DOMElement $element)
+     *
+     * @param Document        $document   DOM document to apply the transformations to.
+     * @param DOMElement      $ampElement Element to adapt.
+     * @param ErrorCollection $errors     Collection of errors that are collected during transformation.
+     * @return bool Whether boilerplate can be removed.
+     */
+    private function adaptBlockingAttributes(Document $document, DOMElement $ampElement, ErrorCollection $errors)
+    {
+        $attributes = $ampElement->attributes;
+
+        if ($attributes === null) {
+            return true;
+        }
+
+        $customCss          = '';
+        $attributesToRemove = [];
+
+        foreach ($attributes as $attribute) {
+            /**
+             * Attribute to check.
+             *
+             * @var DOMAttr $attribute
+             */
+            $normalizedAttributeName = strtolower($attribute->name);
+            if (array_key_exists($normalizedAttributeName, self::EXTRACTION_METHOD_MAP)) {
+                if (
+                    $normalizedAttributeName === Attribute::SIZES
+                    && $ampElement->hasAttribute(Attribute::DISABLE_INLINE_WIDTH)
+                ) {
+                    // Don't remove sizes when disable-inline-width is set.
+                    // @see https://github.com/ampproject/amphtml/pull/27083
+                    continue;
+                }
+
+                $method = self::EXTRACTION_METHOD_MAP[$normalizedAttributeName];
+
+                try {
+                    $customCss            .= $this->$method($document, $ampElement, $attribute->value);
+                    $attributesToRemove[] = $attribute->name;
+                    $ampElement->removeAttribute($attribute->name);
+                } catch (Exception $exception) {
+                    $errors->add(Error\CannotRemoveBoilerplate::fromAttributesRequiringBoilerplate($ampElement));
+                    return false;
+                }
+            }
+        }
+
+        if (!empty($customCss) && ! $this->addCustomCss($document, $customCss)) {
+            $errors->add(Error\CannotRemoveBoilerplate::fromAttributesRequiringBoilerplate($ampElement));
+            return false;
+        }
+
+        foreach ($attributesToRemove as $attributeToRemove) {
+            $ampElement->removeAttribute($attributeToRemove);
+        }
+
+        return true;
+    }
+
+    /**
+     * Add custom CSS styling to the document.
+     *
+     * @param Document $document  Document to add the custom CSS styling to.
+     * @param string   $customCss Custom CSS styling to add.
+     * @return bool Whether adding the custom CSS styling to the document succeeded.
+     */
+    private function addCustomCss(Document $document, $customCss)
+    {
+        $additionalBytes = strlen($customCss);
+
+        if (($this->ampCustomCssByteCount + $additionalBytes) > self::MAX_CSS_BYTE_COUNT) {
+            return false;
+        }
+
+        if (empty($this->ampCustomStyleElement)) {
```

Comment:
Should this check to see if the style element is still in the document (e.g. if it doesn't have a `parentNode`)? If not, then shouldn't `ampCustomCssByteCount` be zeroed out? I realize this is a condition that could never happen given the current transformers, but it's just something that came to mind.

Reply:
I'm not happy with the way the `ampCustomStyleElement` is currently being handled, as it creates a dependency between transformers.

I was thinking of adding an abstraction into `DOM\Document` so that you'll have:
```php
$document->addCustomCss( string $customCss, bool $ignoreByteCount = false );
```

This would work consistently across transformers and add or omit the style element as needed. It would also mean we could count bytes on addition, or even do stuff as automatic minification behind the scenes.

Reply:
Oh, I reread your comment and I think you misread the code here.

The `empty($this->ampCustomStyleElement)` is because we're storing the element as an internal property, however it is not initialized right away.

So the empty check here is just to provide the initialization when we hit `addCustomCss()` for the first time.

Then it tries to retrieve the style element from the actual document to see if one already existed before we ran this transformer, and count the bytes accordingly.

If not (meaning we didn't have a style element AND the document didn't contain one either), we start with a byte count of zero.

Reply:
Yes, my concern/question is if another transformer removed the element after it had been stored. But I suppose that couldn't happen while the transformer is running anyway.